dbcon/execplan/rewrites.cpp

```diff
@@ -0,0 +1,437 @@
+  }
+
+  collectCommonConjuctions(root->left(), accumulator, ++level, orMeeted, operatorType(root) == OP_AND);
+  collectCommonConjuctions(root->right(), accumulator, ++level, orMeeted, operatorType(root) == OP_AND);
+  return;
+}
+
+// this utility function creates new and node
+execplan::ParseTree* newAndNode()
+{
+  execplan::Operator* op = new execplan::Operator();
+  op->data("and");
+  return new execplan::ParseTree(op);
+}
+
+template <typename Common>
+execplan::ParseTree* appendToRoot(execplan::ParseTree* tree, const Common& common)
+{
+  if (common.empty())
+    return tree;
+
+  // TODO: refactor to debug
+  execplan::ParseTree* result = newAndNode();
+  auto current = result;
+  for (auto treenode = common.begin(); treenode != common.end();)
+  {
+    execplan::ParseTree* andCondition = *treenode;
+
+    ++treenode;
+    current->right(andCondition);
+
+    if ((treenode != common.end() && std::next(treenode) != common.end()) ||
+        (std::next(treenode) == common.end() && tree != nullptr))
+    {
+      execplan::ParseTree* andOp = newAndNode();
+      current->left(andOp);
+      current = andOp;
+    }
+    else if (std::next(treenode) == common.end() && tree == nullptr)
+    {
+      current->left(andCondition);
+    }
+  }
+  if (tree)
+    current->left(tree);
+
+  return result;
+}
+
+enum class GoTo
+{
+  Left,
+  Right,
+  Up
+};
+
+struct StackFrame
+{
+  execplan::ParseTree** node;
+  GoTo direction;
+  ChildType containsLeft;
+  ChildType containsRight;
+  StackFrame(execplan::ParseTree** node_, GoTo direction_)
+   : node(node_), direction(direction_), containsLeft(ChildType::Leave), containsRight(ChildType::Leave)
+  {
+  }
+};
+
+using DFSStack = std::vector<StackFrame>;
+
+void deleteOneNode(execplan::ParseTree** node)
+{
+  if (!node || !*node)
+    return;
+
+  (*node)->nullLeft();
+  (*node)->nullRight();
+
+#if debug_rewrites
+  std::cerr << "   Deleting: " << (*node)->data() << " " << boost::core::demangle(typeid(**node).name())
+            << " "
+            << "ptr: " << *node << std::endl;
+#endif
+
+  delete *node;
+  *node = nullptr;
+}
+
+// this utility function adds one stack frame to a stack for dfs traversal
+void addStackFrame(DFSStack& stack, GoTo direction, execplan::ParseTree* node)
+{
+  if (direction == GoTo::Left)
+  {
+    stack.back().direction = GoTo::Right;
+    if (node->left() != nullptr)
+    {
+      auto left = node->leftRef();
+      stack.emplace_back(left, GoTo::Left);
+    }
+  }
+  else if (direction == GoTo::Right)
+  {
+    stack.back().direction = GoTo::Up;
+    if (node->right() != nullptr)
+    {
+      auto right = node->rightRef();
+      stack.emplace_back(right, GoTo::Left);
+    }
+  }
+}
+
+// this utility function reaplces the flag for in the stack frame,
+// indicating whether to delete, unchain or leave child node. It depends on the direction
+// specified in the stack frame
+void replaceContainsTypeFlag(StackFrame& stackframe, ChildType containsflag)
```

Comment:
@mariadb-LeonidFedorov @mariadb-AndreyPiskunov what do you think about to use google code style for the comments in MDB? https://google.github.io/styleguide/cppguide.html#Comment_Style
Currently everyone writes comments in the way it sees, but IMHO comments for the code is very important, as you can check top level google project, they almost have clear comments for every line of the code, I know it could be too much for us, we do not have so much time, but at least we can follow basic grammar rules - like start sentence with a capital letter, end sentence with a dot and so on.

Reply:
Denis, thank you for this notice! Going further I think we owe each other codestyle conversation on CS. Some hot discussion and document to refer. In my perspective, it's effective to use existing codestyle like Google, or LLVM, and refine it with enabling exceptions or specific places for curly braces. I will be glad to make this codestyle rules together. Some of this codestyle conventions are already implicitly implemented. But we still have no hot link to copy-paste on codereview